File: java/src/main/java/benchmarks/BufferHashCRC32.java
package benchmarks;

public class BufferHashCRC32 extends BufferHashBenchmark {

    @Override
    public String name() {
        return "BufferHashCRC32";
    }

    @Override
    long test() {
        int crc = 0xFFFFFFFF;

        for (byte b : data) {
            crc = crc ^ (b & 0xFF);
            for (int j = 0; j < 8; j++) {
                if ((crc & 1) != 0) {
                    crc = (crc >>> 1) ^ 0xEDB88320;
                } else {
                    crc = crc >>> 1;
                }
            }
        }

        return (crc ^ 0xFFFFFFFF) & 0xFFFFFFFFL;
    }
}

File: nim/src/benchmarks/bufferhashcrc32.nim
import ../benchmark
import ../helper
import bufferhash_common

type
  BufferHashCRC32* = ref object of BufferHashBenchmark

proc newBufferHashCRC32(): Benchmark =
  BufferHashCRC32()

method name(self: BufferHashCRC32): string = "BufferHashCRC32"

proc crc32(data: seq[byte]): uint32 =
  var crc = 0xFFFFFFFF'u32

  for byteVal in data:
    crc = crc xor uint32(byteVal)
    for j in 0..<8:
      if (crc and 1) != 0:
        crc = (crc shr 1) xor 0xEDB88320'u32
      else:
        crc = crc shr 1

  crc xor 0xFFFFFFFF'u32

method test(self: BufferHashCRC32): uint32 =
  crc32(self.data)

registerBenchmark("BufferHashCRC32", newBufferHashCRC32)
